src/interface/pipe.rs

```diff
@@ -14,17 +20,42 @@ use std::slice;
 use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
 use std::sync::Arc;
 
+// lets define a few constants for permission flags and the standard size of a Linux page
 const O_RDONLY: i32 = 0o0;
 const O_WRONLY: i32 = 0o1;
 const O_RDWRFLAGS: i32 = 0o3;
 const PAGE_SIZE: usize = 4096;
 
-const CANCEL_CHECK_INTERVAL: usize = 1048576; // check to cancel pipe reads every 2^20 iterations
-
+// lets also define an interval to check for thread cancellations since we may be blocking in trusted space here
+// we're going to define this as 2^20 which should be ~1 sec according to the standard CLOCKS_PER_SEC definition, though it should be quite shorter on modern CPUs
+const CANCEL_CHECK_INTERVAL: usize = 1048576;
+
+/// # Description
+/// Helper function to create pipe objects
+///
+/// # Arguments
+///
+/// * `size` - Size of the iPC construct (either pipe or Unix socket)
+///
+/// # Returns
+///
+/// EmulatedPipe object
+///
 pub fn new_pipe(size: usize) -> EmulatedPipe {
     EmulatedPipe::new_with_capacity(size)
 }
 
+/// # Description
+/// In-memory pipe struct
+///
+/// # Fields
+///
+/// * `write_end` - Reference to the write end of the pipe protected by a RWLock.
+/// * `read_end` - Reference to the read end of the pipe protected by a RWLock.
+/// * `refcount_write` - Count of open write references.
+/// * `refcount_read` - Count of open read references.
+/// * `eof` - Flag signifying the pipe has finished being written to.
+/// * `size` - Size of pipe buffer in bytes.
```

Comment:
Linux pipe implementation does not have eof attributes, whenever it wants to see if the pipe is eof, it just checks if there is still any writers. I think Linux's approach might be better since it does not depend on external caller (in this case, close_syscall) to mark the pipe is in end of file state.

Reply:
Yeah this makes sense, I'm going to try to factor this out.